// @needle-tools/engine/dist/needle-engine.js
import { GameObject as r, onClear as i, onDestroy as a, onInitialized as s, onAfterRender as l, onBeforeRender as d, onUpdate as c, onStart as m, loadAsset as p, NeedleXRSession as g, Context as u, VERSION as S, Component as C, Components as y } from "./needle-engine.bundle-BAsxNKpA.js";
import { $physicsKey as P, ActionBuilder as A, ActionCollection as T, ActionModel as b, Addressables as M, AlignmentConstraint as I, AmbientMode as D, Animation as O, AnimationCurve as x, AnimationExtension as k, AnimationTrackHandler as L, AnimationUtils as N, Animator as F, AnimatorConditionMode as B, AnimatorController as U, AnimatorControllerParameterType as w, AnimatorStateInfo as G, Antialiasing as W, Application as H, AssetDatabase as X, AssetReference as z, Attractor as V, AudioExtension as _, AudioListener as j, AudioSource as K, AudioTrackHandler as Z, Avatar as Q, AvatarBlink_Simple as Y, AvatarEyeLook_Rotation as $, AvatarLoader as J, AvatarMarker as q, AvatarModel as ee, Avatar_Brain_LookAt as te, Avatar_MouthShapes as oe, Avatar_MustacheShake as re, Avatar_POI as ne, Axes as ie, AxesHelper as ae, BUILD_TIME as se, BaseUIComponent as le, BasicIKConstraint as de, BehaviorExtension as ce, BehaviorModel as me, Component$1 as pe, BlobStorage as ge, BloomEffect as ue, BoxCollider as Se, BoxGizmo as Ce, BoxHelperComponent as ye, Button as Ee, ButtonsFactory as Re, CallDirection as he, CallInfo as fe, Camera as ve, CameraTargetReachedEvent as Pe, Canvas as Ae, CanvasGroup as Te, CapsuleCollider as be, ChangeMaterialOnClick as Me, ChangeTransformOnClick as Ie, CharacterController as De, CharacterControllerInput as Oe, ChromaticAberration as xe, CircularBuffer as ke, ClearFlags as Le, ClickThrough as Ne, ClipExtrapolation as Fe, Collider as Be, Collision as Ue, CollisionDetectionMode as we, ColorAdjustments as Ge, ColorBySpeedModule as We, ColorOverLifetimeModule as He, Component$1 as Xe, ComponentLifecycleEvents as ze, ConnectionEvents as Ve, ContactPoint as _e, ContactShadows as je, ContextArgs as Ke, ContextEvent as Ze, ContextRegistry as Qe, ControlTrackHandler as Ye, CursorFollow as $e, CustomBranding as Je, CustomShader as qe, DefaultReflectionMode as et, Deletable as tt, DeleteBox as ot, DepthOfField as rt, DeviceFlag as nt, DeviceUtilities as it, DocumentExtension as at, DragControls as st, DragMode as lt, DropListener as dt, Duplicatable as ct, EffectWrapper as mt, EmissionModule as pt, EmphasizeOnClick as gt, EngineLoadingView as ut, EnvironmentScene as St, EventList as Ct, EventListEvent as yt, EventSystem as Et, EventTrigger as Rt, FieldWithDefault as ht, FileReference as ft, FileReferenceSerializer as vt, FileSpawnModel as Pt, File_Event as At, FixedJoint as Tt, Fog as bt, FrameEvent as Mt, GENERATOR as It, Gizmos as Dt, GltfExport as Ot, GltfExportBox as xt, Gradient as kt, Graphic as Lt, GraphicRaycaster as Nt, Graphics as Ft, GridHelper as Bt, GridLayoutGroup as Ut, GroundProjectedEnv as wt, GroupActionModel as Gt, HideFlags as Wt, HideOnStart as Ht, HingeJoint as Xt, HorizontalLayoutGroup as zt, HostData as Vt, Image as _t, ImageReference as jt, ImageReferenceSerializer as Kt, InheritVelocityModule as Zt, Input as Qt, InputEventQueue as Yt, InputEvents as $t, InputField as Jt, InstanceHandle as qt, InstancingHandler as eo, InstancingUtil as to, InstantiateEvent as oo, InstantiateIdProvider as ro, InstantiateOptions as no, Interactable as io, InternalScreenshotUtils as ao, JoinedRoomResponse as so, KeyEventArgs as lo, Keyframe as co, LODGroup as mo, LODModel as po, LeftRoomResponse as go, Light as uo, LightData as So, LimitVelocityOverLifetimeModule as Co, LoadingElementOptions as yo, LogStats as Eo, LogType as Ro, LookAt as ho, LookAtConstraint as fo, MainModule as vo, MarkerType as Po, MaskableGraphic as Ao, Mathf as To, MeshCollider as bo, MeshRenderer as Mo, MinMaxCurve as Io, MinMaxGradient as Do, NEEDLE_ENGINE_FEATURE_FLAGS as Oo, MODULES as xo, NEKeyboardEvent as ko, NEPointerEvent as Lo, NeedleButtonElement as No, ContextRegistry as Fo, NeedleEngineModelLoader as Bo, NeedleEngineWebComponent as Uo, NeedleMenu as wo, NeedlePatchesKey as Go, USDZExporter$1 as Wo, NeedleXRController as Ho, NeedleXRSync as Xo, NeedleXRUtils as zo, NestedGltf as Vo, NetworkConnection as _o, NetworkedStreamEvents as jo, NetworkedStreams as Ko, Networking as Zo, NewInstanceModel as Qo, NoiseModule as Yo, ObjectRaycaster as $o, ObjectUtils as Jo, OffsetConstraint as qo, OneEuroFilter as er, OneEuroFilterXYZ as tr, OpenURL as or, OrbitControls as rr, Outline as nr, OwnershipEvent as ir, OwnershipModel as ar, PUBLIC_KEY as sr, Padding as lr, ParticleBurst as dr, ParticleSubEmitter as cr, ParticleSystem as mr, ParticleSystemBaseBehaviour as pr, ParticleSystemRenderer as gr, ParticleSystemShapeType as ur, PeerHandle as Sr, PeerNetworking as Cr, Physics as yr, PhysicsExtension as Er, PhysicsMaterialCombine as Rr, PixelationEffect as hr, PlayAnimationOnClick as fr, PlayAudioOnClick as vr, PlayableDirector as Pr, PlayerColor as Ar, PlayerState as Tr, PlayerStateEvent as br, PlayerSync as Mr, PlayerView as Ir, PlayerViewManager as Dr, PointerEventData as Or, PointerType as xr, PostProcessingEffect as kr, PostProcessingEffectOrder as Lr, PostProcessingHandler as Nr, Volume as Fr, PreliminaryAction as Br, PreliminaryTrigger as Ur, PreviewHelper as wr, PrimitiveType as Gr, Progress as Wr, PromiseAllWithErrors as Hr, PromiseErrorResult as Xr, RGBAColor as zr, RapierPhysics as Vr, RawImage as _r, RaycastOptions as jr, Rect as Kr, RectTransform as Zr, ReflectionProbe as Qr, RegisteredAnimationInfo as Yr, RemoteSkybox as $r, RenderTexture as Jr, RenderTextureSerializer as qr, Renderer as en, RendererData as tn, RendererLightmap as on, Rigidbody as rn, RigidbodyConstraints as nn, RoomEvents as an, RotationBySpeedModule as sn, RotationOverLifetimeModule as ln, SceneLightSettings as dn, SceneSwitcher as cn, ScreenCapture as mn, ScreenSpaceAmbientOcclusion as pn, ScreenSpaceAmbientOcclusionN8 as gn, ScrollFollow as un, SendQueue as Sn, SerializationContext as Cn, SetActiveOnClick as yn, ShadowCatcher as En, ShapeModule as Rn, ShapeOverlapResult as hn, SharpeningEffect as fn, SignalAsset as vn, SignalReceiver as Pn, SignalReceiverEvent as An, SignalTrackHandler as Tn, Size as bn, SizeBySpeedModule as Mn, SizeOverLifetimeModule as In, SkinnedMeshRenderer as Dn, SmoothFollow as On, SpatialGrabRaycaster as xn, SpatialHtml as kn, SpatialTrigger as Ln, SpatialTriggerReceiver as Nn, SpectatorCamera as Fn, SphereCollider as Bn, SphereIntersection as Un, SplineContainer as wn, SplineData as Gn, SplineUtils as Wn, SplineWalker as Hn, Sprite as Xn, SpriteData as zn, SpriteRenderer as Vn, SpriteSheet as _n, StateMachineBehaviour as jn, StreamEndedEvent as Kn, StreamReceivedEvent as Zn, SubEmitterSystem as Qn, SyncedCamera as Yn, SyncedRoom as $n, SyncedTransform as Jn, TapGestureTrigger as qn, TeleportTarget as ei, TestRunner as ti, TestSceneUtils as oi, TestSimulateUserData as ri, Text as ni, TextBuilder as ii, TextExtension as ai, TextureSheetAnimationModule as si, TiltShiftEffect as li, Time as di, ToneMappingEffect as ci, TrackHandler as mi, TrackType as pi, TrailModule as gi, TransformData as ui, TransformGizmo as Si, TriggerBuilder as Ci, TriggerModel as yi, TypeStore as Ei, UIRaycastUtils as Ri, UIRootComponent as hi, USDDocument as fi, USDObject as vi, USDWriter as Pi, USDZExporter as Ai, USDZText as Ti, USDZUIExtension as bi, UriSerializer as Mi, UsageMarker as Ii, UserJoinedOrLeftRoomModel as Di, VariantAction as Oi, VelocityOverLifetimeModule as xi, VerticalLayoutGroup as ki, VideoPlayer as Li, ViewDevice as Ni, Vignette as Fi, VisibilityAction as Bi, Voip as Ui, Volume as wi, VolumeParameter as Gi, VolumeProfile as Wi, WaitForFrames as Hi, WaitForPromise as Xi, WaitForSeconds as zi, Watch as Vi, WebARCameraBackground as _i, WebARSessionRoot as ji, WebXR as Ki, WebXRButtonFactory as Zi, WebXRImageTracking as Qi, WebXRImageTrackingModel as Yi, WebXRPlaneTracking as $i, WebXRTrackedImage as Ji, XRControllerFollow as qi, XRControllerModel as ea, XRControllerMovement as ta, XRFlag as oa, XRRig as ra, XRState as na, XRStateFlag as ia, __Ignore as aa, __internalNotifyObjectDestroyed as sa, activeInHierarchyFieldName as la, addAttributeChangeCallback as da, addComponent as ca, addCustomExtensionPlugin as ma, addNewComponent as pa, addPatch as ga, apply as ua, applyHMRChanges as Sa, applyPrototypeExtensions as Ca, beginListenDestroy as ya, beginListenInstantiate as Ea, binaryIdentifierCasts as Ra, build_scene_functions as ha, builtinComponentKeyName as fa, calculateProgress01 as va, clearMessages as Pa, clearMessages as Aa, colorSerializer as Ta, compareAssociation as ba, componentSerializer as Ma, copyTexture as Ia, createMotion as Da, debugNet as Oa, debugOwner as xa, decompressGpuTexture as ka, deepClone as La, delay as Na, delayForFrames as Fa, deserializeObject as Ba, destroy as Ua, destroyComponentInstance as wa, determineMimeTypeFromExtension as Ga, disposeObjectResources as Wa, disposeStream as Ha, editorGuidKeyName as Xa, enableSpatialConsole as za, euler as Va, eventListSerializer as _a, exportAsGLTF as ja, findByGuid as Ka, findObjectOfType as Za, findObjectsOfType as Qa, findResourceUsers as Ya, fitObjectIntoVolume as $a, foreachComponent as Ja, foreachComponentEnumerator as qa, forward as es, generateQRCode as ts, generateSeed as os, getBoundingBox as rs, getCameraController as ns, getComponent as is, getComponentInChildren as as, getComponentInParent as ss, getComponents as ls, getComponentsInChildren as ds, getComponentsInParent as cs, getFormattedDate as ms, getIconElement as ps, getIconTexture as gs, getLoader as us, getOrAddComponent as Ss, getParam as Cs, getParentHierarchyPath as ys, getPath as Es, getPeerOptions as Rs, getPeerjsInstance as hs, getResourceUserCount as fs, getTempColor as vs, getTempQuaternion as Ps, getTempVector as As, getUrlParams as Ts, getVisibleInCustomShadowRendering as bs, getWorldDirection as Ms, getWorldEuler as Is, getWorldPosition as Ds, getWorldQuaternion as Os, getWorldRotation as xs, getWorldScale as ks, hasCommercialLicense as Ls, hasIndieLicense as Ns, hasPointerEventComponent as Fs, hasProLicense as Bs, hideDebugConsole as Us, imageToCanvas as ws, instantiate as Gs, invokeLoadedImportPluginHooks as Ws, invokeXRSessionEnd as Hs, invokeXRSessionStart as Xs, isActiveInHierarchy as zs, isActiveSelf as Vs, isAndroidDevice as _s, isAnimationAction as js, isComponent as Ks, isDebugMode as Zs, isDesktop as Qs, isDestroyed as Ys, isDevEnvironment as $s, isDisposed as Js, isExporting as qs, isGLTFModel as el, isHostedOnGlitch as tl, isHotReloadEnabled as ol, isHotReloading as rl, isIPad as nl, isIconElement as il, isLocalNetwork as al, isMacOS as sl, isMobileDevice as ll, isMozillaXR as dl, isQuest as cl, isResourceTrackingEnabled as ml, isSafari as pl, isUsingInstancing as gl, isiOS as ul, isiPad as Sl, loadPMREM as Cl, loadSync as yl, logHierarchy as El, lookAtInverse as Rl, lookAtObject as hl, lookAtScreenPoint as fl, makeId as vl, makeIdFromRandomWords as Pl, makeNameSafe as Al, markAsInstancedRendered as Tl, microphonePermissionsGranted as bl, nameof as Ml, nameofFactory as Il, objectSerializer as Dl, offXRSessionEnd as Ol, offXRSessionStart as xl, onXRSessionEnd as kl, onXRSessionStart as Ll, parseSync as Nl, placeOnSurface as Fl, postprocessFBXMaterials as Bl, prefix as Ul, pushState as wl, randomNumber as Gl, registerBinaryType as Wl, registerComponent as Hl, registerComponentExtension as Xl, registerCustomEffectType as zl, registerExportExtensions as Vl, registerExtensions as _l, registerHotReloadType as jl, registerLoader as Kl, registerPrefabProvider as Zl, registerPrototypeExtensions as Ql, registerType as Yl, relativePathPrefix as $l, removeAttributeChangeCallback as Jl, removeComponent as ql, removeCustomImportExtensionType as ed, removePatch as td, resolveUrl as od, sanitizeString as rd, saveImage as nd, screenshot as id, screenshot2 as ad, sendDestroyed as sd, serializable as ld, serializeObject as dd, serializeable as cd, setActive as md, setAllowBalloonMessages as pd, setAllowOverlayMessages as gd, setAutoFitEnabled as ud, setCameraController as Sd, setDestroyed as Cd, setDevEnvironment as yd, setDisposable as Ed, setDontDestroy as Rd, setOrAddParamsToUrl as hd, setParam as fd, setParamWithoutReload as vd, setPeerOptions as Pd, setResourceTrackingEnabled as Ad, setState as Td, setVisibleInCustomShadowRendering as bd, setWorldEuler as Md, setWorldPosition as Id, setWorldPositionXYZ as Dd, setWorldQuaternion as Od, setWorldQuaternionXYZW as xd, setWorldRotation as kd, setWorldRotationXYZ as Ld, setWorldScale as Nd, showBalloonError as Fd, showBalloonMessage as Bd, showBalloonWarning as Ud, showDebugConsole as wd, slerp as Gd, syncDestroy as Wd, syncField as Hd, syncInstantiate as Xd, textureToCanvas as zd, tryCastBinary as Vd, tryDetermineMimetypeFromBinary as _d, tryDetermineMimetypeFromURL as jd, tryFindObject as Kd, tryGetGuid as Zd, unregisterHotReloadType as Qd, unwatchWrite as Yd, useForAutoFit as $d, validate as Jd, watchWrite as qd } from "./needle-engine.bundle-BAsxNKpA.js";
import { NEEDLE_progressive as tc } from "./gltf-progressive-egsMzRdv.js";
import { THREE as E } from "./three-CsHK73Zc.js";
const t = {
  VERSION: S,
  Context: u,
  NeedleXRSession: g,
  assets: {
    loadFromURL: p
  },
  onStart: m,
  onUpdate: c,
  onBeforeRender: d,
  onAfterRender: l,
  onInitializedContext: s,
  onDestroyContext: a,
  onClearContext: i
};
globalThis.Needle?.VERSION !== void 0 && console.warn(`Needle Engine is already imported: ${globalThis.Needle.VERSION}`);
function n(e) {
  for (const o in e)
    t[o] = e[o];
}
n(C);
n(y);
for (const e of Object.getOwnPropertyNames(r))
  switch (e) {
    case "prototype":
    case "constructor":
    case "length":
    case "name":
      continue;
    default:
      t[e] = r[e];
      break;
  }
if (!globalThis.Needle)
  globalThis.Needle = t;
else
  for (const e in t)
    globalThis.Needle[e] = t[e];
globalThis.THREE ? console.warn("Three.js is already imported") : globalThis.THREE = E;
export {
  P as $physicsKey,
  A as ActionBuilder,
  T as ActionCollection,
  b as ActionModel,
  M as Addressables,
  I as AlignmentConstraint,
  D as AmbientMode,
  O as Animation,
  x as AnimationCurve,
  k as AnimationExtension,
  L as AnimationTrackHandler,
  N as AnimationUtils,
  F as Animator,
  B as AnimatorConditionMode,
  U as AnimatorController,
  w as AnimatorControllerParameterType,
  G as AnimatorStateInfo,
  W as Antialiasing,
  H as Application,
  X as AssetDatabase,
  z as AssetReference,
  V as Attractor,
  _ as AudioExtension,
  j as AudioListener,
  K as AudioSource,
  Z as AudioTrackHandler,
  Q as Avatar,
  Y as AvatarBlink_Simple,
  $ as AvatarEyeLook_Rotation,
  J as AvatarLoader,
  q as AvatarMarker,
  ee as AvatarModel,
  te as Avatar_Brain_LookAt,
  oe as Avatar_MouthShapes,
  re as Avatar_MustacheShake,
  ne as Avatar_POI,
  ie as Axes,
  ae as AxesHelper,
  se as BUILD_TIME,
  le as BaseUIComponent,
  de as BasicIKConstraint,
  ce as BehaviorExtension,
  me as BehaviorModel,
  pe as Behaviour,
  ge as BlobStorage,
  ue as BloomEffect,
  Se as BoxCollider,
  Ce as BoxGizmo,
  ye as BoxHelperComponent,
  Ee as Button,
  Re as ButtonsFactory,
  he as CallDirection,
  fe as CallInfo,
  ve as Camera,
  Pe as CameraTargetReachedEvent,
  Ae as Canvas,
  Te as CanvasGroup,
  be as CapsuleCollider,
  Me as ChangeMaterialOnClick,
  Ie as ChangeTransformOnClick,
  De as CharacterController,
  Oe as CharacterControllerInput,
  xe as ChromaticAberration,
  ke as CircularBuffer,
  Le as ClearFlags,
  Ne as ClickThrough,
  Fe as ClipExtrapolation,
  Be as Collider,
  Ue as Collision,
  we as CollisionDetectionMode,
  Ge as ColorAdjustments,
  We as ColorBySpeedModule,
  He as ColorOverLifetimeModule,
  Xe as Component,
  ze as ComponentLifecycleEvents,
  Ve as ConnectionEvents,
  _e as ContactPoint,
  je as ContactShadows,
  u as Context,
  Ke as ContextArgs,
  Ze as ContextEvent,
  Qe as ContextRegistry,
  Ye as ControlTrackHandler,
  $e as CursorFollow,
  Je as CustomBranding,
  qe as CustomShader,
  et as DefaultReflectionMode,
  tt as Deletable,
  ot as DeleteBox,
  rt as DepthOfField,
  nt as DeviceFlag,
  it as DeviceUtilities,
  at as DocumentExtension,
  st as DragControls,
  lt as DragMode,
  dt as DropListener,
  ct as Duplicatable,
  mt as EffectWrapper,
  pt as EmissionModule,
  gt as EmphasizeOnClick,
  ut as EngineLoadingView,
  St as EnvironmentScene,
  Ct as EventList,
  yt as EventListEvent,
  Et as EventSystem,
  Rt as EventTrigger,
  ht as FieldWithDefault,
  ft as FileReference,
  vt as FileReferenceSerializer,
  Pt as FileSpawnModel,
  At as File_Event,
  Tt as FixedJoint,
  bt as Fog,
  Mt as FrameEvent,
  It as GENERATOR,
  r as GameObject,
  Dt as Gizmos,
  Ot as GltfExport,
  xt as GltfExportBox,
  kt as Gradient,
  Lt as Graphic,
  Nt as GraphicRaycaster,
  Ft as Graphics,
  Bt as GridHelper,
  Ut as GridLayoutGroup,
  wt as GroundProjectedEnv,
  Gt as GroupActionModel,
  Wt as HideFlags,
  Ht as HideOnStart,
  Xt as HingeJoint,
  zt as HorizontalLayoutGroup,
  Vt as HostData,
  _t as Image,
  jt as ImageReference,
  Kt as ImageReferenceSerializer,
  Zt as InheritVelocityModule,
  Qt as Input,
  Yt as InputEventQueue,
  $t as InputEvents,
  Jt as InputField,
  qt as InstanceHandle,
  eo as InstancingHandler,
  to as InstancingUtil,
  oo as InstantiateEvent,
  ro as InstantiateIdProvider,
  no as InstantiateOptions,
  io as Interactable,
  ao as InternalScreenshotUtils,
  so as JoinedRoomResponse,
  lo as KeyEventArgs,
  co as Keyframe,
  mo as LODGroup,
  po as LODModel,
  go as LeftRoomResponse,
  uo as Light,
  So as LightData,
  Co as LimitVelocityOverLifetimeModule,
  yo as LoadingElementOptions,
  Eo as LogStats,
  Ro as LogType,
  ho as LookAt,
  fo as LookAtConstraint,
  vo as MainModule,
  Po as MarkerType,
  Ao as MaskableGraphic,
  To as Mathf,
  bo as MeshCollider,
  Mo as MeshRenderer,
  Io as MinMaxCurve,
  Do as MinMaxGradient,
  Oo as NEEDLE_ENGINE_FEATURE_FLAGS,
  xo as NEEDLE_ENGINE_MODULES,
  tc as NEEDLE_progressive,
  ko as NEKeyboardEvent,
  Lo as NEPointerEvent,
  No as NeedleButtonElement,
  Fo as NeedleEngine,
  Bo as NeedleEngineModelLoader,
  Uo as NeedleEngineWebComponent,
  wo as NeedleMenu,
  Go as NeedlePatchesKey,
  Wo as NeedleUSDZExporter,
  Ho as NeedleXRController,
  g as NeedleXRSession,
  Xo as NeedleXRSync,
  zo as NeedleXRUtils,
  Vo as NestedGltf,
  _o as NetworkConnection,
  jo as NetworkedStreamEvents,
  Ko as NetworkedStreams,
  Zo as Networking,
  Qo as NewInstanceModel,
  Yo as NoiseModule,
  $o as ObjectRaycaster,
  Jo as ObjectUtils,
  qo as OffsetConstraint,
  er as OneEuroFilter,
  tr as OneEuroFilterXYZ,
  or as OpenURL,
  rr as OrbitControls,
  nr as Outline,
  ir as OwnershipEvent,
  ar as OwnershipModel,
  sr as PUBLIC_KEY,
  lr as Padding,
  dr as ParticleBurst,
  cr as ParticleSubEmitter,
  mr as ParticleSystem,
  pr as ParticleSystemBaseBehaviour,
  gr as ParticleSystemRenderer,
  ur as ParticleSystemShapeType,
  Sr as PeerHandle,
  Cr as PeerNetworking,
  yr as Physics,
  Er as PhysicsExtension,
  Rr as PhysicsMaterialCombine,
  hr as PixelationEffect,
  fr as PlayAnimationOnClick,
  vr as PlayAudioOnClick,
  Pr as PlayableDirector,
  Ar as PlayerColor,
  Tr as PlayerState,
  br as PlayerStateEvent,
  Mr as PlayerSync,
  Ir as PlayerView,
  Dr as PlayerViewManager,
  Or as PointerEventData,
  xr as PointerType,
  kr as PostProcessingEffect,
  Lr as PostProcessingEffectOrder,
  Nr as PostProcessingHandler,
  Fr as PostProcessingManager,
  Br as PreliminaryAction,
  Ur as PreliminaryTrigger,
  wr as PreviewHelper,
  Gr as PrimitiveType,
  Wr as Progress,
  Hr as PromiseAllWithErrors,
  Xr as PromiseErrorResult,
  zr as RGBAColor,
  Vr as RapierPhysics,
  _r as RawImage,
  jr as RaycastOptions,
  Kr as Rect,
  Zr as RectTransform,
  Qr as ReflectionProbe,
  Yr as RegisteredAnimationInfo,
  $r as RemoteSkybox,
  Jr as RenderTexture,
  qr as RenderTextureSerializer,
  en as Renderer,
  tn as RendererData,
  on as RendererLightmap,
  rn as Rigidbody,
  nn as RigidbodyConstraints,
  an as RoomEvents,
  sn as RotationBySpeedModule,
  ln as RotationOverLifetimeModule,
  dn as SceneLightSettings,
  cn as SceneSwitcher,
  mn as ScreenCapture,
  pn as ScreenSpaceAmbientOcclusion,
  gn as ScreenSpaceAmbientOcclusionN8,
  un as ScrollFollow,
  Sn as SendQueue,
  Cn as SerializationContext,
  yn as SetActiveOnClick,
  En as ShadowCatcher,
  Rn as ShapeModule,
  hn as ShapeOverlapResult,
  fn as SharpeningEffect,
  vn as SignalAsset,
  Pn as SignalReceiver,
  An as SignalReceiverEvent,
  Tn as SignalTrackHandler,
  bn as Size,
  Mn as SizeBySpeedModule,
  In as SizeOverLifetimeModule,
  Dn as SkinnedMeshRenderer,
  On as SmoothFollow,
  xn as SpatialGrabRaycaster,
  kn as SpatialHtml,
  Ln as SpatialTrigger,
  Nn as SpatialTriggerReceiver,
  Fn as SpectatorCamera,
  Bn as SphereCollider,
  Un as SphereIntersection,
  wn as SplineContainer,
  Gn as SplineData,
  Wn as SplineUtils,
  Hn as SplineWalker,
  Xn as Sprite,
  zn as SpriteData,
  Vn as SpriteRenderer,
  _n as SpriteSheet,
  jn as StateMachineBehaviour,
  Kn as StreamEndedEvent,
  Zn as StreamReceivedEvent,
  Qn as SubEmitterSystem,
  Yn as SyncedCamera,
  $n as SyncedRoom,
  Jn as SyncedTransform,
  qn as TapGestureTrigger,
  ei as TeleportTarget,
  ti as TestRunner,
  oi as TestSceneUtils,
  ri as TestSimulateUserData,
  ni as Text,
  ii as TextBuilder,
  ai as TextExtension,
  si as TextureSheetAnimationModule,
  li as TiltShiftEffect,
  di as Time,
  ci as ToneMappingEffect,
  mi as TrackHandler,
  pi as TrackType,
  gi as TrailModule,
  ui as TransformData,
  Si as TransformGizmo,
  Ci as TriggerBuilder,
  yi as TriggerModel,
  Ei as TypeStore,
  Ri as UIRaycastUtils,
  hi as UIRootComponent,
  fi as USDDocument,
  vi as USDObject,
  Pi as USDWriter,
  Ai as USDZExporter,
  Ti as USDZText,
  bi as USDZUIExtension,
  Mi as UriSerializer,
  Ii as UsageMarker,
  Di as UserJoinedOrLeftRoomModel,
  S as VERSION,
  Oi as VariantAction,
  xi as VelocityOverLifetimeModule,
  ki as VerticalLayoutGroup,
  Li as VideoPlayer,
  Ni as ViewDevice,
  Fi as Vignette,
  Bi as VisibilityAction,
  Ui as Voip,
  wi as Volume,
  Gi as VolumeParameter,
  Wi as VolumeProfile,
  Hi as WaitForFrames,
  Xi as WaitForPromise,
  zi as WaitForSeconds,
  Vi as Watch,
  _i as WebARCameraBackground,
  ji as WebARSessionRoot,
  Ki as WebXR,
  Zi as WebXRButtonFactory,
  Qi as WebXRImageTracking,
  Yi as WebXRImageTrackingModel,
  $i as WebXRPlaneTracking,
  Ji as WebXRTrackedImage,
  qi as XRControllerFollow,
  ea as XRControllerModel,
  ta as XRControllerMovement,
  oa as XRFlag,
  ra as XRRig,
  na as XRState,
  ia as XRStateFlag,
  aa as __Ignore,
  sa as __internalNotifyObjectDestroyed,
  la as activeInHierarchyFieldName,
  da as addAttributeChangeCallback,
  ca as addComponent,
  ma as addCustomExtensionPlugin,
  pa as addNewComponent,
  ga as addPatch,
  ua as apply,
  Sa as applyHMRChanges,
  Ca as applyPrototypeExtensions,
  ya as beginListenDestroy,
  Ea as beginListenInstantiate,
  Ra as binaryIdentifierCasts,
  ha as build_scene_functions,
  fa as builtinComponentKeyName,
  va as calculateProgress01,
  Pa as clearBalloonMessages,
  Aa as clearOverlayMessages,
  Ta as colorSerializer,
  ba as compareAssociation,
  Ma as componentSerializer,
  Ia as copyTexture,
  Da as createMotion,
  Oa as debugNet,
  xa as debugOwner,
  ka as decompressGpuTexture,
  La as deepClone,
  Na as delay,
  Fa as delayForFrames,
  Ba as deserializeObject,
  Ua as destroy,
  wa as destroyComponentInstance,
  Ga as determineMimeTypeFromExtension,
  Wa as disposeObjectResources,
  Ha as disposeStream,
  Xa as editorGuidKeyName,
  za as enableSpatialConsole,
  Va as euler,
  _a as eventListSerializer,
  ja as exportAsGLTF,
  Ka as findByGuid,
  Za as findObjectOfType,
  Qa as findObjectsOfType,
  Ya as findResourceUsers,
  $a as fitObjectIntoVolume,
  Ja as foreachComponent,
  qa as foreachComponentEnumerator,
  es as forward,
  ts as generateQRCode,
  os as generateSeed,
  rs as getBoundingBox,
  ns as getCameraController,
  is as getComponent,
  as as getComponentInChildren,
  ss as getComponentInParent,
  ls as getComponents,
  ds as getComponentsInChildren,
  cs as getComponentsInParent,
  ms as getFormattedDate,
  ps as getIconElement,
  gs as getIconTexture,
  us as getLoader,
  Ss as getOrAddComponent,
  Cs as getParam,
  ys as getParentHierarchyPath,
  Es as getPath,
  Rs as getPeerOptions,
  hs as getPeerjsInstance,
  fs as getResourceUserCount,
  vs as getTempColor,
  Ps as getTempQuaternion,
  As as getTempVector,
  Ts as getUrlParams,
  bs as getVisibleInCustomShadowRendering,
  Ms as getWorldDirection,
  Is as getWorldEuler,
  Ds as getWorldPosition,
  Os as getWorldQuaternion,
  xs as getWorldRotation,
  ks as getWorldScale,
  Ls as hasCommercialLicense,
  Ns as hasIndieLicense,
  Fs as hasPointerEventComponent,
  Bs as hasProLicense,
  Us as hideDebugConsole,
  ws as imageToCanvas,
  Gs as instantiate,
  Ws as invokeLoadedImportPluginHooks,
  Hs as invokeXRSessionEnd,
  Xs as invokeXRSessionStart,
  zs as isActiveInHierarchy,
  Vs as isActiveSelf,
  _s as isAndroidDevice,
  js as isAnimationAction,
  Ks as isComponent,
  Zs as isDebugMode,
  Qs as isDesktop,
  Ys as isDestroyed,
  $s as isDevEnvironment,
  Js as isDisposed,
  qs as isExporting,
  el as isGLTFModel,
  tl as isHostedOnGlitch,
  ol as isHotReloadEnabled,
  rl as isHotReloading,
  nl as isIPad,
  il as isIconElement,
  al as isLocalNetwork,
  sl as isMacOS,
  ll as isMobileDevice,
  dl as isMozillaXR,
  cl as isQuest,
  ml as isResourceTrackingEnabled,
  pl as isSafari,
  gl as isUsingInstancing,
  ul as isiOS,
  Sl as isiPad,
  p as loadAsset,
  Cl as loadPMREM,
  yl as loadSync,
  El as logHierarchy,
  Rl as lookAtInverse,
  hl as lookAtObject,
  fl as lookAtScreenPoint,
  vl as makeId,
  Pl as makeIdFromRandomWords,
  Al as makeNameSafeForUSD,
  Tl as markAsInstancedRendered,
  bl as microphonePermissionsGranted,
  Ml as nameof,
  Il as nameofFactory,
  Dl as objectSerializer,
  Ol as offXRSessionEnd,
  xl as offXRSessionStart,
  l as onAfterRender,
  d as onBeforeRender,
  i as onClear,
  a as onDestroy,
  s as onInitialized,
  m as onStart,
  c as onUpdate,
  kl as onXRSessionEnd,
  Ll as onXRSessionStart,
  Nl as parseSync,
  Fl as placeOnSurface,
  Bl as postprocessFBXMaterials,
  Ul as prefix,
  wl as pushState,
  Gl as randomNumber,
  Wl as registerBinaryType,
  Hl as registerComponent,
  Xl as registerComponentExtension,
  zl as registerCustomEffectType,
  Vl as registerExportExtensions,
  _l as registerExtensions,
  jl as registerHotReloadType,
  Kl as registerLoader,
  Zl as registerPrefabProvider,
  Ql as registerPrototypeExtensions,
  Yl as registerType,
  $l as relativePathPrefix,
  Jl as removeAttributeChangeCallback,
  ql as removeComponent,
  ed as removeCustomImportExtensionType,
  td as removePatch,
  od as resolveUrl,
  rd as sanitizeString,
  nd as saveImage,
  id as screenshot,
  ad as screenshot2,
  sd as sendDestroyed,
  ld as serializable,
  dd as serializeObject,
  cd as serializeable,
  md as setActive,
  pd as setAllowBalloonMessages,
  gd as setAllowOverlayMessages,
  ud as setAutoFitEnabled,
  Sd as setCameraController,
  Cd as setDestroyed,
  yd as setDevEnvironment,
  Ed as setDisposable,
  Rd as setDontDestroy,
  hd as setOrAddParamsToUrl,
  fd as setParam,
  vd as setParamWithoutReload,
  Pd as setPeerOptions,
  Ad as setResourceTrackingEnabled,
  Td as setState,
  bd as setVisibleInCustomShadowRendering,
  Md as setWorldEuler,
  Id as setWorldPosition,
  Dd as setWorldPositionXYZ,
  Od as setWorldQuaternion,
  xd as setWorldQuaternionXYZW,
  kd as setWorldRotation,
  Ld as setWorldRotationXYZ,
  Nd as setWorldScale,
  Fd as showBalloonError,
  Bd as showBalloonMessage,
  Ud as showBalloonWarning,
  wd as showDebugConsole,
  Gd as slerp,
  Wd as syncDestroy,
  Hd as syncField,
  Xd as syncInstantiate,
  zd as textureToCanvas,
  Vd as tryCastBinary,
  _d as tryDetermineMimetypeFromBinary,
  jd as tryDetermineMimetypeFromURL,
  Kd as tryFindObject,
  Zd as tryGetGuid,
  Qd as unregisterHotReloadType,
  Yd as unwatchWrite,
  $d as useForAutoFit,
  Jd as validate,
  qd as watchWrite
};
